flashinfer/decode.py

```diff
@@ -1971,17 +1975,55 @@ def trtllm_batch_decode_with_kv_cache(
     bmm2_scale: float,  # todo(Yingyi): add dynamic scale tensor later
     window_left: int = -1,
     out: Optional[torch.Tensor] = None,
-) -> torch.Tensor:
+    out_dtype: str = "auto",
+    out_scale_factor: Optional[torch.Tensor] = None,
+    o_sf_scale: Optional[float] = None,
+    o_sf_vec_size: Optional[int] = None,
+) -> Union[torch.Tensor, FP4Tensor]:
     run_func = get_trtllm_gen_fmha_module().trtllm_paged_attention_decode
     sm_count = get_device_sm_count(query.device)
 
-    if out is None:
-        out = torch.empty_like(query)
-    else:
+    if out_dtype == "fp4":
+        assert query.dtype == torch.float8_e4m3fn
+        assert o_sf_scale is not None
+        assert o_sf_vec_size in [None, 16]
+        o_sf_vec_size = o_sf_vec_size or 16
+
+        fp4_out_shape = query.shape[:-1] + (math.ceil(query.shape[-1] / 2),)
+        if out is None:
+            out = torch.empty(fp4_out_shape, dtype=torch.uint8, device=query.device)
+
+        _check_shape_dtype_device(out, fp4_out_shape, torch.uint8, query.device, "out")
+
+        fp4_out_scale_shape = (
+            math.ceil(query.shape[0] / 128) * 128,
+            math.ceil(query.shape[1] * query.shape[2] / o_sf_vec_size / 4) * 4,
+        )
+        if out_scale_factor is None:
+            out_scale_factor = torch.empty(
+                fp4_out_scale_shape, dtype=torch.float8_e4m3fn, device=query.device
+            )
+
+        # Use uint8 as the container dtype to compliant with next fp4 gemm.
+        _check_shape_dtype_device(
+            out_scale_factor,
+            fp4_out_scale_shape,
+            torch.float8_e4m3fn,
+            query.device,
+            "out_scale_factor",
+        )
+    elif out_dtype == "auto":
+        assert out_scale_factor is None
```

Comment:
maybe to tight? if user already provides out/out_scale_factor, we should be able to infer out_dtype from it, right?